downstairs/src/lib.rs

```diff
@@ -5774,4 +5826,230 @@ mod test {
 
         Ok(())
     }
+
+    /*
+     * Test function that will start up a downstairs (at the provided port)
+     * then create a tcp connection to that downstairs, returning the tcp
+     * connection to the caller.
+     */
+    async fn start_ds_and_connect(
+        listen_port: u16,
+        repair_port: u16,
+    ) -> Result<tokio::net::TcpStream> {
+        /*
+         * Pick some small enough values for what we need.
+         */
+        let bs = 512;
+        let es = 4;
+        let ec = 5;
+        let dir = tempdir()?;
+
+        let ads = create_test_downstairs(bs, es, ec, &dir).await?;
+
+        let _ = start_downstairs(
+            ads,
+            "127.0.0.1".parse().unwrap(),
+            None,
+            listen_port,
+            repair_port,
+            None,
+            None,
+            None,
+        )
+        .await
+        .unwrap();
+
+        let sock = TcpSocket::new_v4().unwrap();
+
+        let addr = SocketAddr::new(
+            IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)),
+            listen_port,
+        );
+        Ok(sock.connect(addr).await.unwrap())
+    }
+
+    #[tokio::test]
+    async fn test_version_match() -> Result<()> {
+        // A simple test to verify that sending the current crucible
+        // message version to the downstairs will trigger a response
+        // indicating the version is supported.
+        let tcp = start_ds_and_connect(5555, 5556).await.unwrap();
+        let (read, write) = tcp.into_split();
+        let mut fr = FramedRead::new(read, CrucibleDecoder::new());
+        let mut fw = FramedWrite::new(write, CrucibleEncoder::new());
+
+        let m = Message::HereIAm {
+            version: CRUCIBLE_MESSAGE_VERSION,
+            upstairs_id: Uuid::new_v4(),
+            session_id: Uuid::new_v4(),
+            gen: 1,
+            read_only: false,
+            encrypted: false,
+            supported_versions: Vec::new(),
+        };
+        fw.send(m).await?;
+
+        let f = fr.next().await.unwrap();
+
+        match f {
+            Ok(Message::YesItsMe {
+                version,
+                repair_addr,
+            }) => {
+                assert_eq!(version, CRUCIBLE_MESSAGE_VERSION);
+                assert_eq!(repair_addr, "127.0.0.1:5556".parse().unwrap());
+            }
+            x => {
+                panic!("Invalid answer from downstairs: {:?}", x);
+            }
+        }
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_version_downrev() -> Result<()> {
+        // Test that a newer crucible version will result in a message
+        // indicating their is a version mismatch.
```

Comment:
nit: "there"

Reply:
I know this is the pattern in all the tests, but it seems like the downstairs version is implicit right? (implicitly `CRUCIBLE_MESSAGE_VERSION`?)

Reply:
Yes, the downstairs that is created in `start_ds_and_connect` will be using whatever
the `CRUCIBLE_MESSAGE_VERSION` is.  I'll add a comment to make that clear.

Reply:
Comments added